test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java

```diff
@@ -0,0 +1,273 @@
+    static final int JVMTI_THREAD_STATE_IN_NATIVE                = 0x400000;
+
+    static void runnable() throws Exception {
+        TestStatus status = new TestStatus("JVMTI_THREAD_STATE_RUNNABLE");
+        CountDownLatch ready = new CountDownLatch(1);
+        final boolean[] stopFlag = new boolean[1];
+        Thread vthread = createPinnedVThread(() -> {
+            ready.countDown();
+            int i = 0;
+            while (!stopFlag[0]) {
+                if (i < 200) {
+                    i++;
+                } else {
+                    i = 0;
+                }
+            }
+        });
+        vthread.start();
+        ready.await();
+        testThreadStates(vthread, true, JVMTI_THREAD_STATE_RUNNABLE);
+        stopFlag[0] = true;
+        status.print();
+    }
+
+    static void blockedOnMonitorEnter() throws Exception {
+        // JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER
+        // Thread is waiting to enter a synchronized block/method or,
+        // after an Object.wait(), waiting to re-enter a synchronized block/method.
+        TestStatus status = new TestStatus("JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER");
+        CountDownLatch ready = new CountDownLatch(1);
+        final Object syncObj = new Object();
+        Thread vthread = createPinnedVThread(() -> {
+            ready.countDown();
+            synchronized (syncObj) {
+            }
+        });
+        synchronized (syncObj) {
+            vthread.start();
+            ready.await();
+            Thread.sleep(500); // wait some time to ensure the thread is blocked on monitor
+            testThreadStates(vthread, true, JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER);
+        }
+        status.print();
+    }
+
+    static void waiting(boolean withTimeout) throws Exception {
+        // JVMTI_THREAD_STATE_WAITING
+        // Thread is waiting.
+        // JVMTI_THREAD_STATE_WAITING_INDEFINITELY
+        // Thread is waiting without a timeout. For example, Object.wait().
+        // JVMTI_THREAD_STATE_WAITING_WITH_TIMEOUT
+        // Thread is waiting with a maximum time to wait specified. For example, Object.wait(long).
+        TestStatus status = new TestStatus(withTimeout
+                                           ? ">>JVMTI_THREAD_STATE_WAITING_WITH_TIMEOUT"
+                                           : ">>JVMTI_THREAD_STATE_WAITING_INDEFINITELY");
+        CountDownLatch ready = new CountDownLatch(1);
+        final Object syncObj = new Object();
+        Thread vthread = createPinnedVThread(() -> {
+            synchronized (syncObj) {
+                try {
+                    ready.countDown();
+                    if (withTimeout) {
+                        syncObj.wait(60000);
+                    } else {
+                        syncObj.wait();
+                    }
+                } catch (InterruptedException ex) {
+                    // expected, ignore
+                }
```

Comment:
Shout this code account for spurious wakeups?

Reply:
Fixed